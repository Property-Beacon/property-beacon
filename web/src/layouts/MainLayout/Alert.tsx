import { ReactNode, useEffect } from 'react'
import {
  AiOutlineBell,
  AiOutlineCheckCircle,
  AiOutlineInfoCircle,
  AiOutlineStop,
  AiOutlineWarning
} from 'react-icons/ai'
import { useMainLayoutContext } from './MainLayout'

export type AlertType = {
  level?: 'info' | 'success' | 'warning' | 'error'
  message?: string
  sticky?: boolean
  // Default 5 * 1000 (5s)
  timeout?: number
  // children for actions
  children?: ReactNode
}

const Alert = () => {
  const { alert, setAlert } = useMainLayoutContext()
  const alertLevelCls = alert?.level
    ? alert.level === 'error'
      ? 'alert-error'
      : alert.level === 'info'
      ? 'alert-info'
      : alert.level === 'success'
      ? 'alert-success'
      : alert.level === 'warning'
      ? 'alert-warning'
      : ''
    : ''

  useEffect(
    () => {
      if (alert && !alert.sticky) {
        setTimeout(() => {
          setAlert(null)
        }, alert.timeout ?? 5000)
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [alert]
  )

  return (
    !!alert && (
      <div
        className={`alert ${alertLevelCls} absolute -top-20 mt-2 w-full z-10`}
      >
        <div className="flex-1 place-items-center">
          {alert.level === 'error' ? (
            <AiOutlineStop size={20} className="mr-1" />
          ) : alert.level === 'info' ? (
            <AiOutlineInfoCircle size={20} className="mr-1" />
          ) : alert.level === 'success' ? (
            <AiOutlineCheckCircle size={20} className="mr-1" />
          ) : alert.level === 'warning' ? (
            <AiOutlineWarning size={20} className="mr-1" />
          ) : (
            <AiOutlineBell size={20} className="mr-1" />
          )}
          <span>{alert.message}</span>
        </div>
        <div className="flex-none">{alert.children}</div>
      </div>
    )
  )
}

export default Alert
