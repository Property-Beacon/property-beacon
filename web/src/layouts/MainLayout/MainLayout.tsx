import { useApolloClient } from '@apollo/client'
import { useAuth } from '@redwoodjs/auth'
import { Link, NavLink, routes } from '@redwoodjs/router'
import {
  createContext,
  Dispatch,
  MouseEvent,
  SetStateAction,
  useContext,
  useEffect,
  useState
} from 'react'
import { FiBell, FiLogOut, FiMenu } from 'react-icons/fi'
import { RiArrowDownSLine } from 'react-icons/ri'
import AvatarCell, { QUERY } from 'src/components/AvatarCell'
import type { GetUserById, GetUserByIdVariables } from 'web/types/graphql'
import MainLayoutAlert, { MainLayoutAlertType } from './MainLayoutAlert'

export type MainLayoutContextType = {
  alert?: MainLayoutAlertType
  setAlert?: Dispatch<SetStateAction<MainLayoutAlertType>>
}

const MainLayoutContext = createContext<MainLayoutContextType>({})
const useMainLayoutContext = () => useContext(MainLayoutContext)

const MainLayout: React.FunctionComponent = ({ children }) => {
  const [fullName, setFullName] = useState('')
  const [alert, setAlert] = useState<MainLayoutAlertType>()
  const { watchQuery } = useApolloClient()
  const {
    loading: authorizing,
    logOut,
    currentUser, // Only unique fields are reliable from currentUser
    isAuthenticated
  } = useAuth()

  useEffect(
    () => {
      if (currentUser?.id && !fullName) {
        watchQuery<GetUserById, GetUserByIdVariables>({
          query: QUERY,
          variables: { id: currentUser.id },
          fetchPolicy: 'cache-first',
          nextFetchPolicy: 'cache-first'
        }).subscribe(
          ({
            data: {
              user: { profile }
            }
          }) => {
            if (profile?.fullName) {
              setFullName(profile.fullName)
            }
          }
        )
      }
    },
    // MainLayout is rendered before authentication response, so need to watch currentUser changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [currentUser?.id, fullName]
  )

  const handleLogOut = (e: MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault()
    e.stopPropagation()

    logOut()
      .then()
      .catch((error) => {
        // TODO logging
        console.error(error)
      })
  }

  return (
    <MainLayoutContext.Provider value={{ alert, setAlert }}>
      <div className="p-3">
        <div className="navbar shadow-lg bg-neutral text-base-300 rounded-box py-3 px-4 sm:px-6">
          <Link to={routes.home()} className="h-9" data-testid="navbar-logo">
            <img
              width={36}
              height={36}
              loading="lazy"
              alt="Property Beacon"
              src="/images/icons/icon-128x128.png"
            />
          </Link>
          <div className="flex-1 justify-end">
            <div className="hidden sm:block">
              <Link
                to={routes.home()}
                className="btn btn-ghost btn-sm rounded-btn"
              >
                <span>Blog</span>
              </Link>
              <Link
                to={routes.home()}
                className="btn btn-ghost btn-sm rounded-btn"
              >
                <span>Help</span>
              </Link>
              <Link
                to={routes.home()}
                className="btn btn-ghost btn-sm rounded-btn"
              >
                <span>About</span>
              </Link>
            </div>
            {authorizing ? (
              <button className="btn btn-sm btn-ghost rounded-btn loading text-gray-500">
                Loading
              </button>
            ) : isAuthenticated ? (
              <>
                <NavLink
                  to={routes.dashboard()}
                  activeClassName="text-primary"
                  className="btn btn-ghost btn-sm rounded-btn hidden sm:inline"
                >
                  <span>Dashboard</span>
                </NavLink>
                <div className="flex dropdown dropdown-end">
                  <button className="avatar online mx-2">
                    <FiBell size={22} className="self-center" />
                  </button>
                  <ul className="menu shadow-lg dropdown-content bg-base-100 rounded-box w-80 mt-12 text-neutral">
                    <li className="text-xs">
                      <span>New report from user Brian liu!</span>
                    </li>
                    <li className="text-xs">
                      <span>
                        Ray White has paid the booking fee on{' '}
                        {new Date().toLocaleString()}
                      </span>
                    </li>
                    <li className="text-xs">
                      <span>New signage report from Peter!</span>
                    </li>
                    <li className="text-xs">
                      <span>
                        Ray White has paid their booking fee on{' '}
                        {new Date().toLocaleString()}
                      </span>
                    </li>
                  </ul>
                </div>
                <div className="pl-2 dropdown dropdown-end">
                  <button
                    tabIndex={0}
                    className="flex items-center"
                    data-testid="navbar-user-avatar"
                  >
                    <div className="rounded-full w-8 h-8 text-sm">
                      <AvatarCell id={currentUser?.id} />
                    </div>
                    <RiArrowDownSLine size={20} />
                  </button>
                  <ul
                    data-testid="navbar-menu"
                    className="menu shadow-lg dropdown-content bg-base-100 rounded-box w-60 mt-4 text-neutral"
                  >
                    <li>
                      <NavLink
                        className="flex-col"
                        data-testid="navbar-menu-user"
                        activeClassName="bg-primary-focus"
                        to={routes.settings({ name: 'profile' })}
                      >
                        <div className="flex w-full">
                          <span
                            data-testid="navbar-menu-user-fullname"
                            className="font-bold flex-1"
                          >
                            {fullName || 'n/a'}
                          </span>
                          <span
                            data-testid="navbar-menu-user-role"
                            className="badge badge-accent badge-sm my-auto"
                          >
                            {currentUser?.role || 'n/a'}
                          </span>
                        </div>
                        <div className="flex w-full">
                          <span
                            data-testid="navbar-menu-user-email"
                            className="flex-1 text-sm font-light text-left break-all"
                          >
                            {currentUser?.email || 'n/a'}
                          </span>
                        </div>
                      </NavLink>
                    </li>
                    <li>
                      <a
                        href="/"
                        onClick={handleLogOut}
                        data-testid="navbar-menu-logout"
                      >
                        <FiLogOut size={20} />
                        <span className="ml-2">Logout</span>
                      </a>
                    </li>
                  </ul>
                </div>
              </>
            ) : (
              <>
                <button className="btn btn-sm btn-ghost sm:hidden">
                  <FiMenu size={22} />
                </button>
              </>
            )}
          </div>
        </div>
        <main className="layout-content relative">
          <MainLayoutAlert />
          {children}
        </main>
      </div>
    </MainLayoutContext.Provider>
  )
}

export { MainLayout as default, useMainLayoutContext }
