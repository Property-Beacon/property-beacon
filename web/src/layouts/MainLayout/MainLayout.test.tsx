import { render, screen } from '@redwoodjs/testing'
import MainLayout from './MainLayout'

describe('MainLayout', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<MainLayout />)
    }).not.toThrow()

    const navbarLogo = screen.getByTestId('navbar-logo')

    expect(navbarLogo).toBeVisible()
    expect(navbarLogo).toMatchSnapshot()
    expect(screen.getByText('Loading')).toBeVisible()
  })

  it('renders AvatarCell error fetching user data', async () => {
    const dateStr = new Date().toLocaleString()

    mockCurrentUser({
      id: 'a3f08caa-e057-11eb-ba80-0242ac130004',
      role: 'ADMIN',
      firstName: 'David',
      lastName: 'Smith',
      email: 'example@domain.com',
      logOn: dateStr,
      logOff: dateStr,
      createdAt: dateStr
    })

    mockGraphQLQuery('GetUserById', (_, { ctx }) => {
      // @ts-expect-error Incorrect GraphQLContext type
      ctx.error({ message: 'Error' })

      return null
    })

    render(<MainLayout />)

    expect(await screen.findByTestId('avatar-cell-error')).toBeVisible()
    expect(await screen.findByTestId('navbar-menu')).toBeVisible()
    expect(await screen.findByTestId('navbar-menu-user')).toBeInTheDocument()
    expect(await screen.findByTestId('navbar-menu-logout')).toBeInTheDocument()
    expect(
      await screen.findByTestId('navbar-menu-user-fullname')
    ).toHaveTextContent('n/a')
    expect(
      await screen.findByTestId('navbar-menu-user-role')
    ).toHaveTextContent('ADMIN')
    expect(
      await screen.findByTestId('navbar-menu-user-email')
    ).toHaveTextContent('example@domain.com')
  })

  // TODO: success mockGraphQLQuery test, currently run into data is null unknown issue
})
