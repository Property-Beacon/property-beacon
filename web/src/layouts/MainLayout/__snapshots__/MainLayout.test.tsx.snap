// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout renders successfully 1`] = `
<a
  class="h-9"
  data-testid="navbar-logo"
  href="/"
>
  <img
    alt="Property Beacon"
    height="36"
    loading="lazy"
    src="/images/icons/icon-128x128.png"
    width="36"
  />
</a>
`;
